### DISCUSSION TITLE
rebaseWhen=auto: disable auto rebasing when github merge queue is in use

### DISCUSSION THREAD

**Discussion by rarkins** [renovatebot/renovate#31524] [Suggest an Idea]
### Tell us more.

GitHub's Merge queue already takes care of testing PRs against the head commit of the base branch, so Renovate's automatic rebasing when automerge=true is not necessary.

Merge queue is part of "branch protection" via the UA. e.g. for our main branch:

<img width="799" alt="image" src="https://github.com/user-attachments/assets/b3d8e346-2275-44c8-89e1-56008af4e9a7">

Yet through the branch protections REST API (e.g. `https://api.github.com/repos/renovatebot/renovate/branches/main/protection`) it's not shown.

I also can't find it listed in the REST API, e.g. https://docs.github.com/en/rest/branches/branch-protection?apiVersion=2022-11-28#get-branch-protection and it seems to be missing from GraphQL too: https://docs.github.com/en/graphql/reference/objects#branchprotectionrule

In GraphQL a PR does have a `mergeQueue` and `isMergeQueueEnabled` fields though, although currently we don't fetch PR lists with GraphQL (we generally prefer REST as if-modified-since caching works).

Here's a GraphQL query which seems to work:

```graphql
query MyQuery {
  repository(name: "renovate", owner: "renovatebot") {
    mergeQueue(branch: "main") {
      id
    }
  }
}
```

Result:

```json
{
  "data": {
    "repository": {
      "mergeQueue": {
        "id": "MQ_kwDOBJLedM0dmQ"
      }
    }
  }
}
```

The `id` returns `null` if I test with a dummy branch or on a repo with no queues.

It would be nice if we could get this info as part of the initRepo() graphql call but it seems like it can't be iterated on (i.e. "list all merge queues").

It seems then that this would result in one extra GraphQL API call per base branch per run, when rebaseWhen=auto. However we could potentially cache this information fairly aggressively as it's unlikely to change often and the impact of incorrectly cached info is low.

One other important point: we probably should know if a base branch has merge queue enabled (enforced) because Renovate-native automerging (including for branches) will then stop working.

**Comment by rarkins** [comment]
Related: https://github.com/renovatebot/renovate/issues/31525

**Comment by secustor** [comment]
Also something to consider is that there are not only BranchProtectionRules, but [RepositoryRules](https://docs.github.com/en/enterprise-cloud@latest/repositories/configuring-branches-and-merges-in-your-repository/managing-rulesets/available-rules-for-rulesets#require-merge-queue) too which can require MergeQueues
